### PR TITLE
tools/rbd: expand where option rbd_default_map_options can be set

### PR DESCRIPTION
**Pull request by chrisphoffman**
tools/rbd: expand option rbd_default_map_options
to be set and used at the image and pool level.

Fixes: https://tracker.ceph.com/issues/52850

Signed-off-by: Christopher Hoffman <choffman@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by idryomov — CHANGES_REQUESTED** [review]
Need to cover this in krbd suite.

File: src/tools/rbd/action/Kernel.cc [hunk 1]
```diff
@@ -547,14 +549,76 @@ int execute_map(const po::variables_map &vm,
...
+
+  for (auto& [key, value] : image_map_options) {
+    if (map_options.count(key) == 0) {
+      map_options[key] = value;
+    }
+  }
+
+  for (auto& [key, value] : pool_map_options) {
+    if (map_options.count(key) == 0) {
+      map_options[key] = value;
```
[the remark this answers is not in the document]
Comment (chrisphoffman) [reply src/tools/rbd/action/Kernel.cc:618]: The PR as it stands merges rbd_default_map_options from each layer using precedence. That seems like the unintended functionality and as such will update. I've also included those requirements in the tracker.

File: src/tools/rbd/action/Kernel.cc [hunk 2]
```diff
@@ -547,14 +549,76 @@ int execute_map(const po::variables_map &vm,
...
       &default_map_options);
   if (r < 0) {
     std::cerr << "rbd: couldn't parse default map options" << std::endl;
     return r;
   }
+
+  librados::IoCtx ioctx;
+  librbd::Image image;
+  r = utils::init_and_open_image(pool_name, nspace_name, image_name, "", "",
+                                 false, &rados, &ioctx, &image);
```
[the remark this answers is not in the document]
Comment (chrisphoffman) [reply src/tools/rbd/action/Kernel.cc]: Thanks for the suggestion. This should be addressed. I ended up using init_io_ctx and open_image.

File: qa/rbd/krbd_default_map.t [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  $ ceph config set global rbd_default_map_options $OPTIONS
+  $ ceph osd pool create rbda
+  pool 'rbda' created
+  $ rbd pool init rbda
+  $ rbd create image1 --size 1000 --pool rbda
+  $ DEV=$(rbd map rbda/image1)
+  $ DEV_ID1=$(rbd device list | grep rbda | grep image1 | grep \- | awk '{print $1;}')
+  $ grep -q ${EFFECTIVE} /sys/bus/rbd/devices/$DEV_ID1/config_info
+  $ rbd unmap rbda/image1
+Test at pool level
```
Comment (idryomov) [inline qa/rbd/krbd_default_map.t:13]: Add an empty line before and after these lines to make it easier to navigate between test cases.

**Review by idryomov — CHANGES_REQUESTED** [review]

File: qa/rbd/krbd_default_map.t [hunk 4]
```diff
@@ -0,0 +1,37 @@
+Test at global level
+  $ OPTIONS="crc"
+  $ EFFECTIVE="${OPTIONS}"
```
[the remark this answers is not in the document]
Comment (chrisphoffman) [reply qa/rbd/krbd_default_map.t:3]: Perhaps an `EXPECTED` variable would be better, per the test cases outlined by you below.

File: qa/rbd/krbd_default_map.t (hunk 3 above)
Comment (chrisphoffman) [reply qa/rbd/krbd_default_map.t:31]: This is done.

File: src/tools/rbd/action/Kernel.cc [hunk 5]
```diff
@@ -548,13 +550,64 @@ int execute_map(const po::variables_map &vm,
...
     return r;
   }
+
+  r = utils::open_image(ioctx, image_name, true, &image);
+  if (r < 0) {
+    return r;
+  }
+
+  std::string default_map_value;
+  r = image.metadata_get(METADATA_CONF_PREFIX + default_map_key, &default_map_value);
```
[the remark this answers is not in the document]
Comment (chrisphoffman) [reply src/tools/rbd/action/Kernel.cc]: I did take a look at that. At the time I thought I'd still have to use metadata_get for pool. I wanted to keep things consistent. Upon further inspection, I discovered rbd supports config_list and now both are using that.

File: src/tools/rbd/action/Kernel.cc [hunk 6]
```diff
@@ -548,13 +548,61 @@ int execute_map(const po::variables_map &vm,
...
+
+        if (r < 0) {
+          std::cerr << "rbd: couldn't parse pool map options" << std::endl;
+          return r;
+        }
+
+        break;
+      }
+    }
+  }
```
Comment (idryomov) [inline src/tools/rbd/action/Kernel.cc:594]: What is the purpose of this entire `if` statement?  It just iterates over the same `options` vector...

File: src/tools/rbd/action/Kernel.cc [hunk 7]
```diff
@@ -548,13 +548,61 @@ int execute_map(const po::variables_map &vm,
...
+
+  if (default_map_options.size() == 0) {
+    r = parse_map_options(
+        g_conf().get_val<std::string>(default_map_key),
+        &default_map_options);
```
Comment (idryomov) [inline src/tools/rbd/action/Kernel.cc:599]: This shouldn't be needed.  `image.config_list()` returns the actual (i.e. possibly overridden) values for all options, meaning that the global value should already be accounted for in `options` vector.

File: qa/suites/krbd/basic/tasks/krbd_default_map.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+tasks:
+- cram:
+    clients:
+      client.0:
+      - qa/rbd/krbd_default_map.t
```
Comment (idryomov) [inline qa/suites/krbd/basic/tasks/krbd_default_map.yaml:5]:
```suggestion
      - qa/rbd/krbd_default_map_options.t
```
and the same for the yaml filename.

File: qa/rbd/krbd_default_map.t [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+  $ ceph osd pool create rbda
+  pool 'rbda' created
+  $ rbd pool init rbda
+  $ rbd create rbda/image1 --size 1000
+
+Test at map options level
+
+  $ OPTIONS="alloc_size=65536,lock_on_read"
+  $ EXPECTED="${OPTIONS}"
```
Comment (idryomov) [inline qa/rbd/krbd_default_map.t:11]: Either use braces consistently for all variables or don't use them at all:
```suggestion
  $ EXPECTED="$OPTIONS"
```

File: qa/rbd/krbd_default_map.t [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  $ DEV=$(sudo rbd map rbda/image1 --options $OPTIONS)
+  $ grep -q ${EXPECTED} /sys/bus/rbd/devices/${DEV#/dev/rbd}/config_info
```
Comment (idryomov) [inline qa/rbd/krbd_default_map.t:13]: These greps also need sudo.

**Review by idryomov — CHANGES_REQUESTED** [review]

File: src/tools/rbd/action/Kernel.cc (hunk 6 above)
Comment (chrisphoffman) [reply src/tools/rbd/action/Kernel.cc:594]: Agreed, that needs to be removed.

File: src/tools/rbd/action/Kernel.cc (hunk 7 above)
Comment (chrisphoffman) [reply src/tools/rbd/action/Kernel.cc:599]: I think the issue I've been having is not grasping what you were meaning by image.config_list() pulling in all options. After seeing this message and further testing it clicked for me and I understand now.

File: qa/suites/krbd/basic/tasks/krbd_default_map.yaml (hunk 8 above)
Comment (chrisphoffman) [reply qa/suites/krbd/basic/tasks/krbd_default_map.yaml]: Done.

File: src/tools/rbd/action/Kernel.cc [hunk 11]
```diff
@@ -548,13 +548,34 @@ int execute_map(const po::variables_map &vm,
...
+  if (r < 0) {
+    return r;
+  }
+
+  std::vector<librbd::config_option_t> options;
+  image.config_list(&options);
+  for (const auto &option : options) {
+    if (option.name == "rbd_default_map_options") {
+      r = parse_map_options(option.value, &default_map_options);
+
```
Comment (idryomov) [inline src/tools/rbd/action/Kernel.cc:569]: Drop this empty line.

File: src/tools/rbd/action/Kernel.cc [hunk 12]
```diff
@@ -548,13 +548,34 @@ int execute_map(const po::variables_map &vm,
...
+      if (r < 0) {
+        std::cerr << "rbd: couldn't parse map options" << std::endl;
```
Comment (idryomov) [inline src/tools/rbd/action/Kernel.cc:571]:
```suggestion
        std::cerr << "rbd: couldn't parse default map options" << std::endl;
```

**Review by idryomov — COMMENTED** [review]
Just a couple of nits -- otherwise looks ready to me!

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
jenkins test api

**Comment by idryomov** [comment]
https://pulpito.ceph.com/dis-2022-02-19_01:13:50-krbd:ms_modeless-wip-dis-testing-testing-default-smithi/

**Pull request merged**